core/distributed/partition.cpp

```diff
@@ -0,0 +1,151 @@
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/distributed/partition.hpp>
+
+
+#include "core/distributed/partition_kernels.hpp"
+
+
+namespace gko {
+namespace distributed {
+namespace partition {
+
+
+GKO_REGISTER_OPERATION(count_ranges, partition::count_ranges);
+GKO_REGISTER_OPERATION(build_from_mapping, partition::build_from_mapping);
+GKO_REGISTER_OPERATION(build_from_contiguous, partition::build_from_contiguous);
+GKO_REGISTER_OPERATION(build_ranges_from_global_size,
+                       partition::build_ranges_from_global_size);
+GKO_REGISTER_OPERATION(build_starting_indices,
+                       partition::build_starting_indices);
+GKO_REGISTER_OPERATION(has_ordered_parts, partition::has_ordered_parts);
+
+
+}  // namespace partition
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+std::unique_ptr<Partition<LocalIndexType, GlobalIndexType>>
+Partition<LocalIndexType, GlobalIndexType>::build_from_mapping(
+    std::shared_ptr<const Executor> exec, const Array<comm_index_type>& mapping,
+    comm_index_type num_parts)
+{
+    auto local_mapping = make_temporary_clone(exec, &mapping);
+    size_type num_ranges{};
+    exec->run(partition::make_count_ranges(*local_mapping.get(), num_ranges));
+    auto result = Partition::create(exec, num_parts, num_ranges);
+    GKO_ASSERT(num_ranges == 0 || (num_ranges > 0 && num_parts > 0));
+    if (num_ranges > 0) {
+        exec->run(partition::make_build_from_mapping(
+            *local_mapping.get(), result->offsets_.get_data(),
+            result->part_ids_.get_data()));
+        result->finalize_construction();
+    }
+    return result;
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+std::unique_ptr<Partition<LocalIndexType, GlobalIndexType>>
+Partition<LocalIndexType, GlobalIndexType>::build_from_contiguous(
+    std::shared_ptr<const Executor> exec, const Array<GlobalIndexType>& ranges)
+{
+    auto local_ranges = make_temporary_clone(exec, &ranges);
+    if (ranges.get_num_elems() > 0) {
```

Comment:
If range `i` contains one element, say 3, it would be [3,4), so `ranges[i] == 3` and `ranges[i+1] == 4`. So I guess, indeed, a valid `ranges` must have at least two entries.

Reply:
Sorry about this check, that is some WIP that I forgot to leave out. I will remove that again, and add some validation support when that becomes possible.